pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java

```diff
@@ -0,0 +1,353 @@
+          fileResources.add(metadata.resourceId());
+          ResourceId currentDirectory = metadata.resourceId().getCurrentDirectory();
+          allDirectoryPathSet.add(currentDirectory.toString());
+          // Beam currently creates a temporary empty folder with name '.temp-beam' sometimes under
+          // the directories. These empty subdirectories could not be determined with the existing
+          // Beam file system apis, hence this is hard-coded here so that these empty directories
+          // and their parent directories can be successfully deleted.
+          allDirectoryPathSet.add(
+              currentDirectory
+                  .resolve(".temp-beam", StandardResolveOptions.RESOLVE_DIRECTORY)
+                  .toString());
+        }
+      } else if (matchResult.status() == Status.ERROR) {
+        String errorMessage =
+            String.format("Error matching files under directory %s", rootDirectory.getFilename());
+        logger.error(errorMessage);
+        throw new IOException(errorMessage);
+      }
+    }
+    List<String> allDirectoryPathList = new ArrayList<>(allDirectoryPathSet);
+    allDirectoryPathList.sort(Collections.reverseOrder());
+    List<ResourceId> directoryResourceIdList =
+        allDirectoryPathList.stream()
+            .map(directory -> FileSystems.matchNewResource(directory, true))
+            .collect(Collectors.toList());
+    directoryResourceIdList.add(rootDirectory);
+
+    // Delete the files and directories
+    try {
+      FileSystems.delete(fileResources);
+      FileSystems.delete(directoryResourceIdList);
+    } catch (DirectoryNotEmptyException e) {
+      logger.warn(
+          "Error while deleting a directory, ignoring this and resuming further. error={}",
+          e.getMessage());
+    }
+  }
+
+  private TreeSet<String> getRecentSnapshots(List<ResourceId> paths, int numberOfSnapshotsToReturn)
+      throws IOException {
+    TreeSet<String> treeSet = new TreeSet<>();
+    for (ResourceId resourceId : paths) {
+      // Ignore incomplete DWH snapshots
+      if (!isDWHComplete(resourceId)) {
+        continue;
+      }
+      // Keep only the recent snapshots
+      if (treeSet.size() < numberOfSnapshotsToReturn) {
+        treeSet.add(resourceId.getFilename());
+      } else if (resourceId.getFilename().compareTo(treeSet.first()) > 0) {
+        treeSet.add(resourceId.getFilename());
+        treeSet.pollFirst();
+      }
+    }
+    return treeSet;
+  }
+
+  /**
+   * This method tells us whether the DWH snapshot has been successfully completed by the pipeline.
+   * This is determined by checking the existence of both the timestamp files which needs to be
+   * created at the start and end of the pipeline runs.
+   *
+   * @param dwhResource The dwh resource path which needs to be checked for completeness
+   * @return the status of DWH completeness.
+   * @throws IOException
+   */
+  public boolean isDWHComplete(ResourceId dwhResource) throws IOException {
```

Comment:
nit: In general in names use `Dwh` instead of `DWH` (and for any other abbreviation).

Reply:
Done